Lib/fontbakery/profiles/shaping.py

```diff
@@ -0,0 +1,370 @@
+# Copyright 2020 Google Sans Authors
+# Copyright 2021 Simon Cozens
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import sys
+import textwrap
+from pathlib import Path
+from fontbakery.callable import check, condition
+from fontbakery.checkrunner import FAIL, PASS, SKIP, Section
+from fontbakery.fonts_profile import profile_factory
+from vharfbuzz import Vharfbuzz
+from os.path import basename, relpath
+from stringbrewer import StringBrewer
+from collidoscope import Collidoscope
+
+shaping_basedir = Path("qa", "shaping_tests")
+
+
+profile_imports = ()
+profile = profile_factory(default_section=Section("Shaping Checks"))
+
+PROFILE_CHECKS = [
+    "com.google.fonts/check/shaping/regression",
+    "com.google.fonts/check/shaping/forbidden",
+    "com.google.fonts/check/shaping/collides",
+]
+
+HTML_HEADER = """
+<!doctype html>
+<html>
+    <head>
+        <meta charset="utf-8">
+        <title>Shaping Report</title>
+        <style type="text/css">
+            @font-face {font-family: "TestFont"; src: url(%s);}
+            .tf { font-family: "TestFont"; }
+            html { font-family: sans-serif; }
+            h2 { background: #77f; color: #fafafa; padding: 12px; }
+            pre { font-size: 1.2rem; }
+            li { font-size: 1.2rem; border-top: 1px solid #ddd; padding: 12px; margin-top: 12px; }
+            svg { height: 100px; margin:10px; transform: matrix(1, 0, 0, -1, 0, 0); }
+        </style>
+    </head>
+    <body>
+"""
+
+html_file = None
+
+
+def ensure_html_report_started(vharfbuzz):
```

Comment:
This will explode in multiprocessing mode. I strongly advice moving this into a reporter, reporters are made for this kind of stuff and concurrency is a non-issue there. The data to be reported should be put into an instance of the `Message` class.

We have not yet implement moving generic data via `Message`, but that does not justify this pattern, also the module level global `html_file` is a really bad pattern, not to mention the hard coded path.

Reply:
Yes, the hard coded path is horrible and I would love to avoid it by using a configuration setting or similar. Does fontbakery *have* configuration settings?

Reply:
First, for other reporters, we use command line options, like `--html FILENAME`, `--ghmarkdown FILENAME`, `--json FILENAME` but I agree this doesn't scale well, from `commands/check_profile.py`:

```py
    if args.json:
        import json
        json.dump(sr.getdoc(), args.json, sort_keys=True, indent=4)
        print(f'A report in JSON format has been'
              f' saved to "{args.json.name}"')

    if args.ghmarkdown:
        args.ghmarkdown.write(mdr.get_markdown())
        print(f'A report in GitHub Markdown format which can be useful\n'
              f' for posting issues on a GitHub issue tracker has been\n'
              f' saved to "{args.ghmarkdown.name}"')

    if args.html:
        args.html.write(hr.get_html())
        print(f'A report in HTML format has been saved to "{args.html.name}"')
```

Maybe we could have a more generic option like `--reporter TYPE FILENAME` (or `--reporter TYPE {sub-argument-parser}` if feasible) to make it easier using custom reporters.

> Does fontbakery have configuration settings?

No. I'd like to see a `fontbakery.yaml` file, there are other use cases for it as well, like check result overrides, so we should namespace it.

Reply:
toml, then, or similar.

Reply:
I'd use YAML, better strictYAML. Where's all that TOML hype coming from?